Introduce getClampedRange and getWorldRange

AaronVanGeffen opened the pull request:
The assumption that lead to this bug was that TilePosRangeView would not produce invalid TilePos2 coordinates. For various reasons, it can and should. In short, we have to check that coords are valid before using them.

This PR introduces two new methods, `getClampedRange` and `getWorldRange`, as a means of obtaining a constrained `TilePosRangeView`.

Fixes #2011, a crash when using terraform tools exceeding the map edge

duncanspumpkin commented:
Could you have a quick check of the whole codebase to see if this is done

AaronVanGeffen commented:
Found a few instances of whole-map iteration that wouldn't produce invalid coords, so I've left the check out. However, I also found a few more instances in `World/Industry.cpp` where I thought a guard wouldn't hurt. Pushed a new commit for those.

ZehMatt commented on src/OpenLoco/src/GameCommands/ChangeLandMaterial.cpp:
```diff
@@ -27,6 +27,9 @@ namespace OpenLoco::GameCommands
         World::TilePosRangeView tileLoop{ World::toTileSpace(pointA), World::toTileSpace(pointB) };
         for (const auto& tilePos : tileLoop)
         {
+            if (!World::validCoords(tilePos))
+                continue;
```
I'm honestly not a fan having this in loops, wouldn't it be better to ensure that that the range specified is correct? It's probably better to have a function that clamps TilePosRangeView to the map bounds.

AaronVanGeffen replied:
I agree, this is clearly very error-prone. However, when previously discussed it wasn't an option for `TilePosRangeView` to be constrained in this way? I don't quite recall. I would prefer the range view to handle this as well, though.

duncanspumpkin replied:
Yeah that could work

AaronVanGeffen replied:
I've pushed a suggestion introducing `getClampedRange` and `getWorldRange` as a means of obtaining a `TilePosRangeView`. Still needs testing.

AaronVanGeffen commented on src/OpenLoco/src/Map/TileManager.cpp:
```diff
@@ -746,10 +746,8 @@ namespace OpenLoco::World::TileManager
+        auto range = getClampedRange(initialTilePos - TilePos2{ 5, 5 }, initialTilePos + TilePos2{ 5, 5 });
         for (auto& tilePos : range)
         {
             if (!World::validCoords(tilePos))
```
Can be removed.

AaronVanGeffen commented on src/OpenLoco/src/World/Industry.cpp:
```diff
@@ -569,6 +570,10 @@ namespace OpenLoco
         std::size_t i = 0;
         for (const auto& tilePos : TilePosRangeView{ topRight, bottomLeft })
         {
+            if (!World::validCoords(tilePos))
```
Can be removed.

ZehMatt commented on src/OpenLoco/src/GameCommands/ChangeLandMaterial.cpp:
```diff
@@ -24,7 +24,7 @@ namespace OpenLoco::GameCommands
         }
 
-        World::TilePosRangeView tileLoop{ World::toTileSpace(pointA), World::toTileSpace(pointB) };
+        auto tileLoop = World::getClampedRange(pointA, pointB);
```
Can be also const

ZehMatt commented on src/OpenLoco/src/GameCommands/ClearLand.cpp:
```diff
@@ -60,7 +60,7 @@ namespace OpenLoco::GameCommands
     static uint32_t clearLand(const ClearLandArgs& args, const uint8_t flags)
     {
-        World::TilePosRangeView tileLoop{ World::toTileSpace(args.pointA), World::toTileSpace(args.pointB) };
+        auto tileLoop = World::getClampedRange(args.pointA, args.pointB);
```
Can be const

ZehMatt reviewed (commented):
Tested it a bit and looks like nothing broke to the best of my knowledge. Some of the variables can be made const otherwise this seems quite nice.

duncanspumpkin commented on src/Engine/include/OpenLoco/Engine/World.hpp:
```diff
@@ -88,4 +89,14 @@ namespace OpenLoco::World
         return Pos2{ static_cast<coord_t>(coords.x * kTileSize), static_cast<coord_t>(coords.y * kTileSize) };
     }
+
+    constexpr coord_t clampCoord(coord_t coord)
```
I'm not sure these are quite correct. We have basically two possible range limits. drawable and normal. drawable is 1 tile less than max/min and normal is max/min. This looks to be a bit of a mix of both. We should also probably define a drawable max min values as they are used a few times now.

ZehMatt replied:
You are right but I don't think this is used anywhere in the drawing code at the moment, I also mentioned the oddity to @AaronVanGeffen on Discord that we should probably have separate constants for the world space as the +1, -1 offset seems quite odd to have for when we should rather have proper constants.

AaronVanGeffen replied:
Any suggestions for what those constants should look like?

ZehMatt replied:
I think the code is fine as is, accessing tiles at 0,0 or the max is an invalid memory read. The drawing code interprets coordinates out of that range as void. We can refine the constants at a later point to better separate whats used for rendering and whats used for accessing the tile memory. @duncanspumpkin what are your thoughts?

duncanspumpkin replied:
there are also tile manager functions that use the drawableCoords validator so i think we do need this to be correct for the various setups.

ZehMatt replied:
Not sure I understand what you mean, when it comes to getting a Tile then it must be within the `1, 1` and `kkMapWidth, kMapColumns`, I don't see any code modified that isn't about getting memory for the tiles currently.

duncanspumpkin replied:
I'm concerned about minor divergences. At present for example look at the audio change this was previously ranged to max less< kMapColumns now it is to 1-> kMapColumns-1. The upper range is the same as it's gone from < to <=. The lower range has truncated by 1. <0 we can ignore as it would crash before. And that's just for the tile clamp. For the world clamp it's also had its lower bound truncated from 0 to 1. Which doesn't match a drawable coord (32) or a world coord (0). The upper bound is still ultimately the same though as before.

So either the clamps should be 0->Columns-1 or they should be 1->Columns-2 and for world coords 0->Width-1 or 32->Width-33.

ZehMatt replied:
Hm perhaps we should have two distinct functions then, getWorldRange and getDrawableRange and hide the clamping functions. Given the new functions the code shouldn't have to manually clamp coordinates anymore.

LeftofZen replied:
is this resolved?

AaronVanGeffen replied:
I believe I've now adjusted the clamps as suggested.

ZehMatt commented:
Here is what I think. Rename and move TileLoop.cpp to Engine/World.cpp and drop the clamp functions from World.hpp and move the one to clamp tile coords also in World.cpp, this function should not be directly used as thats what the new functions are for, I don't think we'll ever need to just clamp one axis.

AaronVanGeffen commented:
Sounds good to me. I'll be afk for most of the evening (on mobile now), but if you like you can push any changes you see fit.

duncanspumpkin commented on src/OpenLoco/src/Audio/Audio.cpp:
```diff
@@ -844,16 +844,13 @@ namespace OpenLoco::Audio
             const auto topLeft = World::toTileSpace(centre) - World::TilePos2{ 5, 5 };
             const auto bottomRight = topLeft + World::TilePos2{ 11, 11 };
-            World::TilePosRangeView searchRange(topLeft, bottomRight);
+            const auto searchRange = World::getClampedRange(topLeft, bottomRight);
```
Following notes are just for me to confirm I've looked at the loop:
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/GameCommands/ChangeLandMaterial.cpp:
```diff
@@ -24,7 +24,7 @@ namespace OpenLoco::GameCommands
         }
 
-        World::TilePosRangeView tileLoop{ World::toTileSpace(pointA), World::toTileSpace(pointB) };
+        const auto tileLoop = World::getClampedRange(pointA, pointB);
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/GameCommands/ClearLand.cpp:
```diff
@@ -60,7 +60,7 @@ namespace OpenLoco::GameCommands
     static uint32_t clearLand(const ClearLandArgs& args, const uint8_t flags)
     {
-        World::TilePosRangeView tileLoop{ World::toTileSpace(args.pointA), World::toTileSpace(args.pointB) };
+        const auto tileLoop = World::getClampedRange(args.pointA, args.pointB);
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/GameCommands/LowerLand.cpp:
```diff
@@ -33,7 +33,7 @@ namespace OpenLoco::GameCommands
         }
 
-        World::TilePosRangeView tileLoop{ toTileSpace(args.pointA), toTileSpace(args.pointB) };
+        const auto tileLoop = World::getClampedRange(args.pointA, args.pointB);
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/GameCommands/LowerWater.cpp:
```diff
@@ -28,7 +28,7 @@ namespace OpenLoco::GameCommands
         }
 
-        World::TilePosRangeView tileLoop{ toTileSpace(args.pointA), toTileSpace(args.pointB) };
+        const auto tileLoop = getClampedRange(args.pointA, args.pointB);
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/GameCommands/RaiseLand.cpp:
```diff
@@ -33,7 +33,7 @@ namespace OpenLoco::GameCommands
         }
 
-        World::TilePosRangeView tileLoop{ toTileSpace(args.pointA), toTileSpace(args.pointB) };
+        const auto tileLoop = getClampedRange(args.pointA, args.pointB);
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/GameCommands/RaiseWater.cpp:
```diff
@@ -28,7 +28,7 @@ namespace OpenLoco::GameCommands
         }
 
-        World::TilePosRangeView tileLoop{ toTileSpace(args.pointA), toTileSpace(args.pointB) };
+        const auto tileLoop = getClampedRange(args.pointA, args.pointB);
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/GameCommands/RemoveIndustry.cpp:
```diff
@@ -93,7 +93,7 @@ namespace OpenLoco::GameCommands
     static void revokeAllSurfaceClaims(const IndustryId id)
     {
-        for (auto& pos : World::TilePosRangeView({ 1, 1 }, { World::kMapRows - 1, World::kMapColumns - 1 }))
+        for (auto& pos : World::getWorldRange())
```
Vanilla < 0x2FFF x/y started at 0,0

duncanspumpkin commented on src/OpenLoco/src/GameCommands/RemoveTown.cpp:
```diff
@@ -44,9 +44,7 @@ namespace OpenLoco::GameCommands
-        TilePosRangeView tileLoop{ { 1, 1 }, { kMapColumns - 1, kMapRows - 1 } };
-
-        for (auto& tilePos : tileLoop)
+        for (auto& tilePos : getWorldRange())
```
Vanilla < 0x2FFF x/y started at 0,0

duncanspumpkin commented on src/OpenLoco/src/Map/IndustryElement.cpp:
```diff
@@ -217,14 +217,9 @@ namespace OpenLoco::World
-            const auto tileStart = World::toTileSpace(loc);
-            for (auto& tilePos : TilePosRangeView(tileStart - World::TilePos2{ kLowerRange, kLowerRange }, tileStart + World::TilePos2{ upperRange, upperRange }))
+            const auto tileStart = toTileSpace(loc);
+            for (auto& tilePos : getClampedRange(tileStart - TilePos2{ kLowerRange, kLowerRange }, tileStart + TilePos2{ upperRange, upperRange }))
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/Map/MapGenerator.cpp:
```diff
@@ -583,7 +583,7 @@ namespace OpenLoco::World::MapGenerator
         }
 
-        TilePosRangeView tileLoop{ { 1, 1 }, { kMapColumns - 1, kMapRows - 1 } };
+        const TilePosRangeView tileLoop = getWorldRange();
```
Vanilla < 0x2FE0 x/y started at 0x20,0x20

duncanspumpkin commented on src/OpenLoco/src/Map/MapGenerator.cpp:
```diff
@@ -658,7 +658,7 @@ namespace OpenLoco::World::MapGenerator
         uint32_t i = 0;
         std::vector<TileElement*> toBeRemoved;
-        for (auto& loc : TilePosRangeView({}, { kMapRows - 1, kMapColumns - 1 }))
+        for (auto& loc : getWorldRange())
```
Vanilla < 0x2FFF x/y started at 0,0

duncanspumpkin commented on src/OpenLoco/src/Map/TileManager.cpp:
```diff
@@ -746,16 +746,10 @@ namespace OpenLoco::World::TileManager
-            initialTilePos - World::TilePos2{ 5, 5 }, initialTilePos + World::TilePos2{ 5, 5 }
-        };
+        const auto initialTilePos = toTileSpace(loc);
+        auto range = getClampedRange(initialTilePos - TilePos2{ 5, 5 }, initialTilePos + TilePos2{ 5, 5 });
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/Map/TileManager.cpp:
```diff
@@ -804,15 +798,12 @@ namespace OpenLoco::World::TileManager
         // Search a 10x10 area centred at pos.
         // Initial tile position is the top left of the area.
-        auto initialTilePos = World::toTileSpace(pos) - World::TilePos2(5, 5);
+        auto initialTilePos = toTileSpace(pos) - TilePos2(5, 5);
```
Vanilla <= 0x2FFF x/y (didn't have a < 0 check)

duncanspumpkin commented on src/OpenLoco/src/World/Industry.cpp:
```diff
@@ -530,10 +530,11 @@ namespace OpenLoco
         const auto bottomLeft = initialTilePos + TilePos2{ 2, 2 };
-        for (const auto& tilePos : TilePosRangeView{ topRight, bottomLeft })
+
+        for (const auto& tilePos : getClampedRange(topRight, bottomLeft))
```
Vanilla no checks!

duncanspumpkin commented on src/OpenLoco/src/World/Industry.cpp:
```diff
@@ -567,7 +568,7 @@ namespace OpenLoco
 
         std::size_t i = 0;
-        for (const auto& tilePos : TilePosRangeView{ topRight, bottomLeft })
+        for (const auto& tilePos : getClampedRange(topRight, bottomLeft))
```
Vanilla No checks!

duncanspumpkin commented on src/OpenLoco/src/World/TownManager.cpp:
```diff
@@ -80,8 +80,7 @@ namespace OpenLoco::TownManager
 
-        World::TilePosRangeView tileLoop{ { 1, 1 }, { World::kMapColumns - 1, World::kMapRows - 1 } };
-        for (const auto& tilePos : tileLoop)
+        for (const auto& tilePos : World::getWorldRange())
```
Vanilla start at 0,0 to <= 0x2FFF

duncanspumpkin reviewed (commented):
I've added a note as to what vanilla did at every place we've added the clamp

AaronVanGeffen commented:
> I've added a note as to what vanilla did at every place we've added the clamp

Thanks. I think we shouldn't skip the lower bound check where vanilla didn't bother to check it, though. Especially when networking gets involved, I really want to make sure we don't open these up to extra attack vectors.

duncanspumpkin commented:
> > I've added a note as to what vanilla did at every place we've added the clamp
> 
> Thanks. I think we shouldn't skip the lower bound check where vanilla didn't bother to check it, though. Especially when networking gets involved, I really want to make sure we don't open these up to extra attack vectors.

Yes i agree with you on that. I think we just need to ensure that there are only 2: the 0x2FE0 > / <= 32 (draw) and; the 0x2FFF >= / <= 0 (world);. And that the correct one is being used each time. Also note my >< symbols. 0x2FFF is a valid world but 0x2FFE0 is not a valid draw the first valid draw would be 0x2FDF.

AaronVanGeffen commented:
> Yes i agree with you on that. I think we just need to ensure that there are only 2: the 0x2FE0 > / 32 <= (draw) and; the 0x2FFF >= / 0 <= (world);. And that the correct one is being used each time. Also note my >< symbols. 0x2FFF is a valid world but 0x2FFE0 is not a valid draw the first valid draw would be 0x2FDF.

I see what you mean. Let's introduce not just `getWorldRange`, but a `getDrawableTileRange` as well then. On it.

AaronVanGeffen commented:
I have introduced a getDrawableTileRange (`TilePosRangeView({ 1, 1 }, { kMapColumns - 1, kMapRows - 1 })`) to distinguish from getWorldRange (now `TilePosRangeView({ 0, 0 }, { kMapColumns, kMapRows })`). However, this appears to have reintroduced the crashes this PR sought to resolve. Moreover, using terraform tools at map edges are now taking the invisible tiles into account, leading to 'Too low!' errors.

Initially, I had also adjusted clampCoord to from `std::clamp<coord_t>(coord, 1, kMapWidth - 1)` to `std::clamp<coord_t>(coord, 0, kMapWidth)` as well, but that immediately led to crashes. So I have left getClampedRange untouched.

LeftofZen commented on CHANGELOG.md:
```diff
@@ -1,6 +1,8 @@
 23.06.1+ (???)
 ------------------------------------------------------------------------
+- Feature: [#2013] Add headers for base/clearance height, direction, and ghost flag to the tile inspector.
```
i'm guessing this was just forgotten from #2013?

AaronVanGeffen replied:
That's correct.

LeftofZen reviewed (approved):
there's a couple of unresolved comments/conversations, @duncanspumpkin can you confirm if your comments have been fixed?

otherwise, playtested and this is working for me, very nice change with the warning message about off-map terrain adjustments now, rather than current master which is just the terrain adjustment sound with no action and no warning

AaronVanGeffen commented:
In the latest commits, I have introduced a getDrawableTileRange (`TilePosRangeView({ 1, 1 }, { kMapColumns - 2, kMapRows - 2 })`) to distinguish from getWorldRange (now `TilePosRangeView({ 0, 0 }, { kMapColumns - 1, kMapRows - 1 })`). Using terraform tools at map edges are now taking the invisible tiles into account, though, leading to 'Too low!' errors. This appears to match vanilla behaviour, though -- but it looks like vanilla would just silently exit the terraform functions instead.

I think we should consider adjusting the `getClampedRange` function instead such that these invisible tiles are not taken into account again. In an earlier state of this PR, that made the terraform tools more convenient to use.
